src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/13.sql

```diff
@@ -0,0 +1,3443 @@
+) ON [PRIMARY]
+GO
+
+ALTER TABLE dbo.ResourceChangeData ADD  CONSTRAINT DF_ResourceChangeData_Timestamp  DEFAULT (sysutcdatetime()) FOR Timestamp
+GO
+
+/*************************************************************
+    Resource change type table
+**************************************************************/
+CREATE TABLE dbo.ResourceChangeType(
+    ResourceChangeTypeId smallint IDENTITY(0,1) NOT NULL,
+    Name nvarchar(50) NOT NULL,
+ CONSTRAINT PK_ResourceChangeType PRIMARY KEY CLUSTERED 
+(
+    ResourceChangeTypeId ASC
+)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
+) ON [PRIMARY]
+GO
+
+SET IDENTITY_INSERT [dbo].[ResourceChangeType] ON 
+
+INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (0, N'Creation')
+INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (1, N'Update')
+INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (2, N'Deletion')
+
+SET IDENTITY_INSERT dbo.ResourceChangeType OFF
+GO
+
+
+/*************************************************************
+    Stored procedures for capturing and fetching resource changes
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     CaptureResourceChanges
+--
+-- DESCRIPTION
+--     Inserts resource change data
+--
+-- PARAMETERS
+--     @isDeleted
+--         * Whether this resource marks the resource as deleted.
+--     @version
+--         * The version of the resource being written
+--     @resourceId
+--         * The resource ID
+--     @resourceTypeId
+--         * The ID of the resource type
+--
+-- RETURN VALUE
+--     It does not return a value.
+--
+CREATE PROCEDURE dbo.CaptureResourceChanges
+    @isDeleted bit, 
+    @version int, 
+    @resourceId varchar(64),
+    @resourceTypeId smallint
+AS
+BEGIN
+    DECLARE @changeType SMALLINT
+    IF (@isDeleted = 1) BEGIN
+        SET @changeType = 2    /* DELETION */
+    END
+    ELSE BEGIN
+        IF (@version = 1) BEGIN
+            SET @changeType = 0 /* CREATION */
+        END
+        ELSE BEGIN
+            SET @changeType = 1 /* UPDATE */
+        END
+    END
+
+    INSERT INTO dbo.ResourceChangeData
+        (ResourceId, ResourceTypeId, ResourceVersion, ResourceChangeTypeId)
+    VALUES
+        (@resourceId, @resourceTypeId, @version, @changeType)
+END
+GO
+
+--
+-- STORED PROCEDURE
+--     FetchResourceChanges
+--
+-- DESCRIPTION
+--     Returns the number of resource change records from startId
+--
+-- PARAMETERS
+--     @startId
+--         * The start id of resource change records to fetch.
+--     @pageSize
+--         * The page size for fetching resource change records.
+--
+-- RETURN VALUE
+--     Resource change data rows.
+--
+CREATE PROCEDURE dbo.FetchResourceChanges
+    @startId bigint, 
+    @pageSize int
+AS
+BEGIN
+
+    SET NOCOUNT ON;
+
+    SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
```

Comment:
Hi @brendankowitz, please be aware of this change here. 

we have this ResourceChangeData table that the FHIR server will write into and the Event Agent will read from. With this change, when this select query is performed, it will block other writes, currently the upserting sproc. We have a few backup plans to make it better for synchronization and concurrency handling, but those will need more cycles to prove and implement. 

Currently, the impact of this select should be so small, we project 200 items limit per query and 1 query per second, this literally should be pretty minimum impact, except some severe cases maybe like SQL connection issues. 

Also, the EventAgent should be the only consumer of this at this point, and EventAgent will only be turned on for customers who opt-in the event Private Preview. Before Public Preview, we will re-evaluate this part and review with FHIR team again.